quic-client/src/lib.rs

```diff
@@ -0,0 +1,231 @@
+            .get(index)
+            .cloned()
+            .ok_or(ConnectionPoolError::IndexOutOfRange)
+    }
+
+    fn create_pool_entry(
+        &self,
+        config: &Self::TpuConfig,
+        addr: &SocketAddr,
+    ) -> Self::PoolTpuConnection {
+        Quic(Arc::new(QuicClient::new(
+            self.endpoint.clone(),
+            *addr,
+            config.compute_max_parallel_streams(),
+        )))
+    }
+}
+
+pub struct QuicConfig {
+    client_certificate: Arc<QuicClientCertificate>,
+    maybe_staked_nodes: Option<Arc<RwLock<StakedNodes>>>,
+    maybe_client_pubkey: Option<Pubkey>,
+}
+
+impl Default for QuicConfig {
+    fn default() -> Self {
+        let (certs, priv_key) = new_self_signed_tls_certificate_chain(
+            &Keypair::new(),
+            IpAddr::V4(Ipv4Addr::new(0, 0, 0, 0)),
+        )
+        .expect("Failed to initialize QUIC client certificates");
+        Self {
+            client_certificate: Arc::new(QuicClientCertificate {
+                certificates: certs,
+                key: priv_key,
+            }),
+            maybe_staked_nodes: None,
+            maybe_client_pubkey: None,
+        }
+    }
+}
+
+impl QuicConfig {
+    fn create_endpoint(&self) -> Arc<QuicLazyInitializedEndpoint> {
+        Arc::new(QuicLazyInitializedEndpoint::new(
+            self.client_certificate.clone(),
+        ))
+    }
+
+    fn compute_max_parallel_streams(&self) -> usize {
+        let (client_type, stake, total_stake) =
+            self.maybe_client_pubkey
+                .map_or((ConnectionPeerType::Unstaked, 0, 0), |pubkey| {
+                    self.maybe_staked_nodes.as_ref().map_or(
+                        (ConnectionPeerType::Unstaked, 0, 0),
+                        |stakes| {
+                            let rstakes = stakes.read().unwrap();
+                            rstakes.pubkey_stake_map.get(&pubkey).map_or(
+                                (ConnectionPeerType::Unstaked, 0, rstakes.total_stake),
+                                |stake| (ConnectionPeerType::Staked, *stake, rstakes.total_stake),
+                            )
+                        },
+                    )
+                });
+        compute_max_allowed_uni_streams(client_type, stake, total_stake)
+    }
+
+    pub fn update_client_certificate(
+        &mut self,
+        keypair: &Keypair,
+        ipaddr: IpAddr,
+    ) -> Result<(), Box<dyn Error>> {
+        let (certs, priv_key) = new_self_signed_tls_certificate_chain(keypair, ipaddr)?;
+        self.client_certificate = Arc::new(QuicClientCertificate {
+            certificates: certs,
+            key: priv_key,
+        });
+        Ok(())
+    }
+
+    pub fn set_staked_nodes(
+        &mut self,
+        staked_nodes: &Arc<RwLock<StakedNodes>>,
+        client_pubkey: &Pubkey,
+    ) {
+        self.maybe_staked_nodes = Some(staked_nodes.clone());
+        self.maybe_client_pubkey = Some(*client_pubkey);
+    }
+}
+
+pub struct Quic(Arc<QuicClient>);
```

Comment:
I don't really like that this is basically just a wrapper on top of QuicClient. We basically of a Vec of Arc<Arc<QuicClient>>... Would it be possible to just impl BaseTpuConnection for QuicClient or something?

Reply:
The struct just replaces the wrapping `BaseTpuConnection` enum. Are you saying you think we can remove the Arc wrapper for the pool elements?

Reply:
Like perhaps we could just have QuicPool::connections be Vec\<Arc\<QuicClient\>\> and just impl BaseTpuConnection for QuicClient, since currently the struct Quic is just an Arc on a QuicClient. This means that QuicPool::connections is effectively a Vec\<Arc\<Arc\<QuicClient\>\>\>

Reply:
Pretty minor aesthetic nitpick... just really don't like unnecessary Arc derefs, but probably not a huge deal perf-wise (and might even get optimized away anyway)

Reply:
There are Arcs sprayed pretty liberally around ConnectionCache. I'd definitely be open to follow-up work to simplify them in TpuConnectionCache, if you'd like to chip in.